Update maintainer gudelines (#568)
microsoft/vcpkg-docs: vcpkg/contributing/maintainer-guide.md

---
# Maintainer guide

This document describes policies, guidelines, and best practices to follow when making contributions to vcpkg.

It is intended to serve the same role as:

- [Debian's Policy Manual](https://www.debian.org/doc/debian-policy/),
- [Homebrew's Maintainer Guidelines](https://docs.brew.sh/Maintainer-Guidelines), and
- [Homebrew's Formula Cookbook](https://docs.brew.sh/Formula-Cookbook).

## Curated registry design goals

### Ports must install simultaneously

A port included in the curated registry must build successfuly in the presence of all the other ports in the registry.
The entire combination of ports is tested regularly by vcpkg's automated CI.

A port must not do any of the following:

- [Install files that conflict with another port.](#unique-port-attribution-rule)
- Install symbols and definitions owned by another package.
- Test for the presence or absence of another port. Except for declaring a dependency.

Exceptions are made for:

- Ports that replace another one in an official capacity. A successor project.
- Ports where there is no platform overlap. For example, a Windows-only replacement to a Linux project that doesn't
  support Windows.

> - Ports that precede this policy and have a demostrably significant amount of users and contributors.

Author: ```suggestion
- Ports that precede this policy and have a demonstrably significant number of users and contributors.
```

Ports that don't conform to this policy can't be accepted into the curated registry.

[!INCLUDE [registry-or-overlay](../../includes/registry-or-overlay.md)]

### Ports must be tested in at least one official triplet

All ports in the curated registry must be tested in CI for at least one official [triplet](../concepts/triplets.md).

Exceptions are made for:

- Temporary circumstances, such as:
  - The port doesn't build without an update to the CI infrastructure.
  - The port doesn't build because of a recent breaking change in an upstream dependency and there is work in progress
    to resolve the situation.

> - Ports that precede this policy and have a demostrably significant amount of users and contributors.

Author: ```suggestion
- Ports that precede this policy and have a demonstrably significant number of users and contributors.
```

  For example, a new Boost library or Qt component.
- The project demonstrates equivalent maturity to the previous requirements in some other capacity.

Some indicators of project immaturity are:

- The project doesn't show up in search engines.
- The project has frequent renames.
- Conflicts with other libraries.

[!INCLUDE [registry-or-overlay](../../includes/registry-or-overlay.md)]

## PR structure

### Make separate pull requests per port

Smaller pull requests (PRs) are easier to review. Pull requests should make changes to a single port. This also reduces
the wait times for Continuous Integration (CI) results.

Limiting PRs to a single port may not be possible in some cases. For example, when changing a port requires downstream
consumers to be updated or patched.

### Avoid trivial changes in untouched files

Avoid unnecessary trivial changes to a port, such as: reformatting, renaming variables, or fixing typos. If these changes
are not made as part of a necessary modification. Any change that doesn't affect the output of a port's installation is
considered trivial. Trivial changes consume CI resources that are better utilized otherwise.

> It's acceptable to make these type of changes as part of a more significant change, for example, as part of a port update.

Reviewer: I think the previous wording did a better job of conveying "don't create PRs whose whole purpose is these trivial changes, but those changes are otherwise OK. The previous version was also shorter.

How about something like:

```suggestion
Don't make trivial changes to an otherwise unmodified port, such as: reformatting, renaming variables, or fixing typos. Any change that doesn't affect the output of a port's installation is
considered trivial. Trivial changes consume compute time better utilized otherwise.
```

#### Close inactive PRs

The vcpkg team may close pull requests (PRs) that have no activity for more than 60 days.

For pull requests in a reviewable state, the countdown begins from the last time a vcpkg maintainer makes a request for
changes, or asks for clarification. If no activity is made by the contributor within 60 days, the PR is considered stale
and may be closed at the team's discretion.

A PR is in a reviewable state when it:

- Passes all CI tests.

>   In case that a CI test fails, the contributor asks for help or justifies the reason for the failures.

Reviewer: Unfortunately this one is hard these days because GitHub won't run some Actions by default without our action. Also, using "CI" here creates confusion between "PR" builds and "CI" builds; we don't run "CI" against PRs.

How about:

```suggestion
- Has no failing PR checks, or the contributor asks for help or justifies the reason for the failures.
```

- Doesn't have any pending requested changes or clarifications from a vcpkg maintainer.
- Doesn't have merge conflicts.

> - Isn't a draft.

Reviewer: I don't think we would close something which was otherwise OK just because it happens to be a draft?

```suggestion
```

Ideally, portfiles should be short, simple, and as declarative as possible.
Remove any boiler plate comments introduced by the `create` command before submitting a PR.

> ### Use lowercase for hexadecimal digits strings

Author: This was on the "Design Goals" section, which is obviously the wrong place. I moved it to the "Portfiles" section.
